tests/unit/s2n_async_pkey_test.c

```diff
@@ -296,6 +341,48 @@ int main(int argc, char **argv)
             EXPECT_SUCCESS(s2n_config_free(server_config));
             EXPECT_SUCCESS(s2n_config_free(client_config));
         }
+
+        /*  Test: sign with different pkey and apply with s2n_connection having different chain_and_key */
+        {
+
+            struct s2n_config *server_config, *client_config;
+            EXPECT_NOT_NULL(server_config = s2n_config_new());
+            EXPECT_SUCCESS(s2n_config_set_async_pkey_validation_mode(server_config, S2N_ASYNC_PKEY_VALIDATION_STRICT));
+            EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key_to_store(server_config, chain_and_key));
+            EXPECT_SUCCESS(s2n_config_add_dhparams(server_config, dhparams_pem));
+            EXPECT_SUCCESS(s2n_config_set_async_pkey_callback(server_config, async_pkey_store_callback));
+            server_config->security_policy = &server_security_policy;
+
+            EXPECT_NOT_NULL(client_config = s2n_config_new());
+            EXPECT_SUCCESS(s2n_config_set_unsafe_for_testing(client_config));
+
+            EXPECT_SUCCESS(s2n_config_set_verification_ca_location(client_config, S2N_DEFAULT_TEST_CERT_CHAIN, NULL));
+
+            /* Create connection */
+            struct s2n_connection *client_conn = s2n_connection_new(S2N_CLIENT);
+            EXPECT_NOT_NULL(client_conn);
+            EXPECT_SUCCESS(s2n_connection_set_config(client_conn, client_config));
+
+            struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER);
+            EXPECT_NOT_NULL(server_conn);
+            EXPECT_SUCCESS(s2n_connection_set_config(server_conn, server_config));
+
+            /* Create nonblocking pipes */
+            struct s2n_test_io_pair io_pair;
+            EXPECT_SUCCESS(s2n_io_pair_init_non_blocking(&io_pair));
+            EXPECT_SUCCESS(s2n_connection_set_io_pair(client_conn, &io_pair));
+            EXPECT_SUCCESS(s2n_connection_set_io_pair(server_conn, &io_pair));
+
+            EXPECT_EQUAL(try_handshake(server_conn, client_conn, async_handler_sign_with_different_pkey_and_apply),
+                    S2N_FAILURE);
```

Comment:
This is what EXPECT_FAILURE / EXPECT_FAILURE_WITH_ERRNO are for. And how do you know the handshake is failing for the right reason, and not because you set up the test wrong?

Reply:
So sign operation will return `S2N_ERR_VERIFY_SIGNATURE`, which i checked during handshake call using `EXPECT_FAILURE_WITH_ERRNO`
But I'm explicitly failing decrypt operator with S2N_FAILURE, as the current test cases iterate for all ciphers 

To sum it up, i mentioned `S2N_FAILURE` in the main. And catching right error no. during sign operation

Reply:
I think I'm following. I don't see the point of returning S2N_FAILURE without setting a useful errno you can check for, because otherwise this test could be failing anywhere.

Why not make your callback succeed, and check for success?

Reply:
Done